src/coreclr/jit/morph.cpp

```diff
@@ -9241,7 +9241,7 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
                 noway_assert(op1);
                 if (op1->IsCnsIntOrI())
                 {
-                    noway_assert(varTypeIsIntOrI(tree));
+                    noway_assert(varTypeIsIntegralOrI(tree));
```

Comment:
It may still do a wrong thing if op2 is a non-zero byref, I don't think we currently support those, but nothing stops us from adding those I think, for frozen objects

Reply:
although, I expect many similar foldings will fail too